[examples/summarization] deal with `max_length` and `num_beams`

bofenghuang opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Hello 👋,

1. In the examples/pytorch/summarization/run_summarization.py, when the `generation_max_length` is not defined, the parameter `max_length` for `generate` function will be set to `val_max_target_length` in the line [675](https://github.com/huggingface/transformers/blob/82e61f34451dbea2de8d2220d51b0609d605dfd7/examples/pytorch/summarization/run_summarization.py#L675-L679), and be used for the **final evaluation** and prediction after training. However, the `max_length` used for the **evaluation during the training** will be possibly set to `None` or the `max_length` defined in the model's config. It's not a consistent behavior. Here I tried to unify this parameter before the `Seq2SeqTrainer` is initialized. Also applied the same to `num_beams`.

2. In the examples/pytorch/summarization/run_summarization_no_trainer.py, fixed the parameter `max_length`. Think it has higher priority than `val_max_target_length`.

Please ignore this PR if it's intended behavior :)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc: @sgugger @patil-suraj 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer: @stas00, Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

bofenghuang committed:
Override the decoding parameters of Seq2SeqTrainer

bofenghuang committed:
Fix quality

bofenghuang committed:
Fix max_length parameter

bofenghuang committed:
Fix quality

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_21740). All of your documentation changes will be reflected on that endpoint.

sgugger commented on examples/pytorch/summarization/run_summarization_no_trainer.py:
```diff
@@ -164,11 +164,8 @@ def parse_args():
-            "The maximum total input sequence length after tokenization. Sequences longer than this will be truncated,"
-            " sequences shorter will be padded if `--pad_to_max_lengh` is passed."
-        ),
+        default=None,
+        help=("The maximum target length to use when predicting with the generate method."),
```
This is the maximum length for the inputs, not the targets. I don't think this should be touched.

bofenghuang replied:
There is another argument for the inputs `max_source_length`. So I suppose the `max_length` is used to constrain the generation length

sgugger replied:
It's actually never used if I look more closely, so let's remove it entirely.

sgugger commented on examples/pytorch/summarization/run_summarization_no_trainer.py:
```diff
@@ -671,7 +668,7 @@ def postprocess_text(preds, labels):
 
         gen_kwargs = {
-            "max_length": args.val_max_target_length if args is not None else config.max_length,
+            "max_length": args.max_length if args.max_length is not None else args.val_max_target_length,
```
It should be the `max_target_length` here if `val_max_target_length` is `None`, `max_length` is for the inputs only.

sgugger replied:
So you can just use `max_target_length` here or the config max_length if it was set to None.

bofenghuang replied:
`max_target_length` is also used in `preprocess_function` to truncate the outputs. 

Here I think it might be better to provide another generation max length parameter, which is only used for evaluation/prediction, as done in examples/pytorch/summarization/run_summarization.py. In case the users don't want to truncate the summaries in dataset, but want to limit the generation max length in evaluation to accelerate the whole training process?

And of course if it is `None`, it will be set to the `val_max_target_length` (or `max_target_length`) or the `max_length` in model's config

sgugger replied:
I don't think so. The parameter comes from a copy-paste and wasn't used, I wouldn't find a use for it. If the user wants some none-standard behavior, they should adapt the example.

sgugger reviewed (commented):
Thanks for your PR, I left a few comments.

bofenghuang committed:
Remove redundant parameter max_length

bofenghuang committed:
Separate the preprocess of train and validation to use different max_target_length

bofenghuang commented:
Hi @sgugger, just removed the `max_length`!

I also added another modif which separates the preprocessing of the training and the validation set in order to truncate the outputs by `max_target_length` and `val_max_target_length` respectively

sgugger reviewed (approved):
Thanks for iterating!
